### ISSUE TITLE
Change datastructure proposal

### ISSUE THREAD

**Issue by videlec** [videlec/sage-flatsurf#27]
This is a proposal to change the datastructures we currently use. The main concern is that it is easier to work with oriented edges as building block rather than polygons. This is how [flipper](https://github.com/MarkCBell/flipper), [veerer](https://gitlab.com/videlec/veerer/) or [libflatsurf](https://github.com/flatsurf/flatsurf) work.

A surface would be encoded by a set of labeled "half-edges". In flipper labels are `L = {-n, ..., 0, 1, ..., n-1}` while in libflatsurf this is `L = {-n, ..., -1, 1, ..., n}`. The CW complex structure is defined via
- an involution on `L` telling which half-edge is glued to which half-edge. In flipper the involution is fixed to be bit complement `~` (`0 <-> -1`, `1 <-> -2`) while in libflatsuf it is the negation `-`.
- a permutation `next_in_face` on `L` (whose cycles describe the faces)
- a permutation `next_at_vertex` on `L` (whose cycles describe the vertices)
This triple of permutations defines the topology of the surface. To define a geometric structure, one needs additional data describing the monodromy. For translation surfaces this would simply be a vector for each half-edge such that `monodromy(~e) = - monomdromy(e)`.

To build the square torus
```
sage: TranslationSurface(faces='(0,1,~0,~1)', holonomies=[(1,0), (0,1)])
```

Such datastructure is also convenient to construct slit surfaces. One defines two surfaces with boundaries `S1` and `S2` where the labels are disjoint and the boundary are exchanged by the involution. Then take the union.

**Comment by wphooper** [comment]
First a few questions:

Are you using the terms monodromy and holonomy to mean the same thing? In your square torus example, I guess because it is a translation surface you are implicilty extending the definition of holonomies (currently define for indices 0 and 1) such that holonomies[~n]=-holonomies[n], correct?

For half-translation surfaces, I guess you would specify holonomy as a dictionary which would be required to satisfy holonomies[~n]= \pm holonomies[n]. For similarity surfaces that condition would be completely absent, correct?

In our current setup, polygons get labels. It seems like the suggestion is either not to have polygon labels or to deemphasize them. This seems like it will break virtually everything.

Perhaps the basic data structure we use should be a _subsurface_ with some indexing scheme involving a finite subset of natural numbers. A subsurface would then keep track of a finite collection of unglued edges (empty in the case of a closed surface). Subsurfaces seem superior to surfaces because:
* Operations (e.g., triangle flips, slit constructions, rel deformations, etc) can be performed easily: pull polygons out and then replace them. The data type would remain the same even if we perform complicated operations. This seems intuitive.
* Infinite surfaces can be thought of as finite subsurfaces together with a function which extends across any unglued edge.

I want the code to be able to interact easily with libflatsurf, and I expect we can agree on a better redesign for the underlying data structures. Perhaps we should declare a stable branch (declare the most recent working version as "1.1") and start a development branch working towards 2.0. In the development branch we can basically start over (and then port over code we care about). Again, I think we will break virtually everything, and just trying to do it with the code base we have will be a nightmare.

**Comment by videlec** [comment]
You are correct: we want to specify holonomies (I will update the description).

Operations like forgetting/adding/flipping an edge is expensive when the data structure is polygon centered (you have to destruct two polygons and rebuild two). When it is edge centered this is a very simple update of the permutation and holonomy (a bit more care when it is not translation).

It is possible and definitely desirable to support face or vertex labels. This can be easily implemented as follows. A labelling of the faces is simply a map `L -> FaceLabels` which is invariant under `next_in_face` (idem for vertex label `L -> VertexLabels` invariant under `next_at_vertex`). This labelling is also simple to update when performing flips.

I am not so much worried about breaking the code base. What I want to include with this proposal is a step by step move forward

1. design something independent of the old code base to handle the half edge centered datastructure
2. use it as a "backend" in the old code base
3. move more and more of the old code base to the backend

**Comment by wphooper** [comment]
I'm okay with trying this. Can you tag the current master as "0.2" before we start making these changes?

**Comment by videlec** [comment]
As soon as I start, I will!
